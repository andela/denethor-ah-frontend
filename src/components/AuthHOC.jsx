import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

const authHOC = (AuthencticatedComponent) => {
  class AuthHOC extends Component {
    componentDidUpdate () {
      if (this.props.isLoggedIn === false) {
        this.props.history.push('/login')
      }
    }

    render() {
      return <AuthencticatedComponent {...this.props}/>
    }
  }

  AuthHOC.propTypes = {
    isLoggedIn: PropTypes.bool,
    history: PropTypes.object
  }

  return AuthHOC;
}

const ConnectedAuthHOC = (Component) => {
  const AuthComponent = authHOC(Component);
  const mapStateToProps = ({ auth: { isLoggedIn } }) => ({ isLoggedIn });
  return connect(mapStateToProps)(AuthComponent);
}

export default ConnectedAuthHOC;