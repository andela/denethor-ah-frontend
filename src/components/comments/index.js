import CommentEntries from './CommentEntries';
import CreateComment from './CreateComment';

export {CommentEntries, CreateComment};