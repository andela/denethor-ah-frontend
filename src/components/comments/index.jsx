import CommentEntries from './CommentEntries';

export { CommentEntries };