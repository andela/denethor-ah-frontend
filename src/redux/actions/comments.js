import axios from '../../utils/axiosConfig';
import {
  ADD_COMMENT,
  REMOVE_COMMENT,
  GET_COMMENT_SUCCESS
} from './types';

const BASE_URL = process.env.API_ROOT_URL;

export const addCommentRequest = () => ({
  type: ADD_COMMENT
});

export const getCommentSuccess = comments => ({
  type: GET_COMMENT_SUCCESS,
  comments
});

export const getArticleComments = (articleId) => async (dispatch) => {
    const { data: { articleComments } } = await axios.get(`${BASE_URL}/articles/${articleId}/comments`);
    dispatch(getCommentSuccess(articleComments));
};
export const addComment = ({ articleId, commentBody }) => async (dispatch) => {
  try {
    dispatch(addCommentRequest());
    await axios.post(`${BASE_URL}/articles/${articleId}/comments`, { commentBody });
    dispatch(getArticleComments(articleId));
  } catch (error) {
    throw error;
  }
};


export const removeComment = id => (dispatch) => {
  dispatch({
    type: REMOVE_COMMENT,
    id
  });
};