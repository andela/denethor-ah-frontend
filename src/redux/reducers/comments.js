import {
  ADD_COMMENT,
  REMOVE_COMMENT,
  GET_COMMENT_SUCCESS
} from '../actions/types';

export const commentReducerDefaultState = {
  comments: []
};

export default (state = commentReducerDefaultState, action) => {
  switch (action.type) {
    case ADD_COMMENT:
      return {
        ...state
      }
    case GET_COMMENT_SUCCESS:
      return {
        ...state,
        comments: [...action.comments]
      };

    case REMOVE_COMMENT:
      return state.filter(({
        id
      }) => id !== action.id);

    default:
      return state;
  }
};
