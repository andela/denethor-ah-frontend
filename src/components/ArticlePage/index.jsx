import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import RoundedImage from '../../components/RoundedImage';
import VerticalMargin from '../../components/VerticalMargin';
import HorizontalMargin from '../../components/HorizontalMargin';
import RatingStars from '../../components/RatingStars';
import RatingStarsBox from '../../components/RatingStarsBox';
import HorizontalLine from '../../components/HorizontalLine';
import TextAreaInput from '../../components/TextAreaInput';
import CommentEntries from '../../components/CommentEntries';
import { getOneArticle, rateArticle, getArticleAvgRating } from '../../redux/actions/articles';
import { toast } from 'react-toastify';
import './style.scss';


export class SingleArticleView extends Component {

  async componentDidMount() {
    const { fetchArticle, getArticleAvgRating, match } = this.props;
    let articleId = match.params.articleId;
    await fetchArticle(articleId);
    getArticleAvgRating(articleId)
  }

  starClickHandle = (item) => {
    let articleId = this.props.match.params.articleId;
    this.props.rateArticle(item, articleId )
    .then(() => toast.success('Thank you for rating this article'))
    .catch(error => {
      if (error.response.data.message === 'You already rated this article'){
        return toast.error('You have already rated this article');
      } else if(error.response.data === 'Unauthorized'){
         toast.error('You need to login to rate an article');
        return  this.props.history.push('/login')
      }
      toast.error('Error while rating article');
    })
  }

  render() {
    const { match: { params: { articleId } }, articles = [] } = this.props;
    const article = articles.find(item => item.id === articleId) || {};
    const { 
      title, 
      featuredImage = '/assets/img/map-typing.jpg', 
      author: { username = '', imageUrl: userAvatarImage = '' } = {}, 
      body,
      averageRating,
      ratingsCount,
      comments = [],
    } = article;

    return (
      <div>
        <div className='banner-section'>
          <div className='banner-image'>
            <img src={featuredImage} alt='banner background' />
          </div>
          <div className='transparent-overlay' />
          <div className='banner-content'>
            <div className='auto-container'>
              <div className='banner-content-title'>{title}</div>
              <div className='avatar-text-group'>
                <div className='user-profile-picture'>
                  <RoundedImage imageSource={userAvatarImage} alt='profile picture'/>
                  <VerticalMargin className='hide-for-medium' size={50} />
                </div>
                <HorizontalMargin className='show-for-medium' size={10} />
                <div className='article-username-rate-group'>
                  <div className='article-header-username'>{username}</div>
                  <div className ='article-rating-container'>
                    <div className='article-header-rating'>
                      <RatingStars rateNumber={(averageRating)} />
                    </div>
                    <div className='article-header-rated-times'>
                      {!ratingsCount ? 'No ratings yet' : `rated by ${ratingsCount} user${ratingsCount > 1 ? 's' : ''}`}
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div className='body-section'>
          <div className='auto-container wide-padding'>
            <div className='article-body'>
              <div dangerouslySetInnerHTML={{ __html: body }} />
            </div>
            <HorizontalLine />
            <div className='section-article-info'>  
              <div>Rate this Article <RatingStarsBox starClickHandle= {this.starClickHandle.bind(this)} /></div>  
              <VerticalMargin className='show-for-medium' size={10} />
              <div>
                <div className='section-comment-create'>
                  <div>Leave comments</div>
                  <TextAreaInput className='article-comment-field' text='Your message' />
                  <button></button>
                </div>
                <div className='section-user-comments'>
                  <CommentEntries comments={comments} />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

SingleArticleView.propTypes = {
  articles: PropTypes.array.isRequired,
  match: PropTypes.object.isRequired,
  fetchArticle: PropTypes.func.isRequired,
  rateArticle: PropTypes.func.isRequired,
  getArticleAvgRating: PropTypes.func.isRequired,
  history: PropTypes.object
}

const mapStateToProps = (state) => ({
  articles: state.articles
});

const mapDispatchToProps = (dispatch) => ({
  fetchArticle: (id) => dispatch(getOneArticle(id)),
  rateArticle: (rating, articleId) => dispatch(rateArticle(rating, articleId)),
  getArticleAvgRating: (articleId) => dispatch(getArticleAvgRating(articleId)),
});

export default connect(mapStateToProps, mapDispatchToProps)(SingleArticleView);



