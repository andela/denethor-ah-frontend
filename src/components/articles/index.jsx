import ArticlePage from './ArticlePage';
import ArticleCreatePage from './articleCreate/ArticleCreate';

export { ArticlePage, ArticleCreatePage };