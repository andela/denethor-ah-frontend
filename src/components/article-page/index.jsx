import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import RoundedImage from '../../components/RoundedImage';
import VerticalMargin from '../../components/VerticalMargin';
import HorizontalMargin from '../../components/HorizontalMargin';
import RatingStars from '../../components/RatingStars';
import RatingStarsBox from '../../components/RatingStarsBox';
import HorizontalLine from '../../components/horizontal-line';
import TextAreaInput from '../../components/TextAreaInput';
import CommentEntries from '../../components/CommentEntries';
import { getOneArticle, rateArticle, getArticleAvgRating, likeArticle, dislikeArticle, bookmarkArticle } from '../../redux/actions/articles';
import { toast } from 'react-toastify';
import './style.scss';
import { LikeNumberGroup , BookMarkNumberGroup, DisLikeNumberGroup } from '../like-number-group';
// import impressionsControl from '../../utils/impressionScroll';



export class SingleArticleView extends Component {

  async componentDidMount() {
    const { fetchArticle, getArticleAvgRating, match } = this.props;
    let articleId = match.params.articleId;
    await fetchArticle(articleId);
    getArticleAvgRating(articleId)
  }

  starClickHandle = (item) => {
    let articleId = this.props.match.params.articleId;
    this.props.rateArticle(item, articleId )
    .then(() => toast.success('Thank you for rating this article'))
    .catch(error => {
      if (error.response.data.message === 'You already rated this article'){
        return toast.error('You have already rated this article');
      } else if(error.response.data === 'Unauthorized'){
         toast.error('You need to login to rate an article');
        return  this.props.history.push('/login')
      }
      toast.error('Error while rating article');
    })
  }

  likeHandle = () => {
    let articleId = this.props.match.params.articleId;
    this.props.likeArticle(articleId)
    .then((response) => {
      const message = response.data.data.message;
      toast.success(`${message}`)
    })
    .catch(error => {
      if(error.response === 'Unauthorized'){
        return  toast.error('You need to login to like an article');
     }
       toast.error(`${error}`);
    })
  }

  dislikeHandle = () => {
    let articleId = this.props.match.params.articleId;
    this.props.dislikeArticle(articleId)
    .then((response) => {
      const message = response.data.data.message;
      toast.success(`${message}`)
    })
    .catch(error => {
      if(error.response === 'Unauthorized'){
        return toast.error('You need to login to dislike an article');
     }
      return toast.error('An error occurred while trying to dislike the article');
    })
  }

  bookmarkHandle = () => {
    let articleId = this.props.match.params.articleId;
    this.props.bookmarkArticle(articleId)
    .then(() => toast.success('Thank you for bookmarking this article'))
    .catch(error => {
      if(error.response === 'Unauthorized'){
        return toast.error('You need to login to bookmark an article');      
     } else if(error.response.status === 409){
       return toast.error('You have already bookmarked this article')
     }
      return toast.error('An error occurred while trying to bookmark this article');
    })
  }

  render() {
    const { match: { params: { articleId } }, articles = [] } = this.props;
    const article = articles.find(item => item.id === articleId) || {};
    const { 
      title, 
      featuredImage = '/assets/img/map-typing.jpg', 
      author: { username = '', imageUrl: userAvatarImage = '' } = {}, 
      body,
      averageRating,
      ratingsCount,
      comments = [],
    } = article;

    return (
      <div>
        <div className='banner-section'>
          <div className='banner-image'>
            <img src={featuredImage} alt='banner background' />
          </div>
          <div className='transparent-overlay' />
          <div className='banner-content'>
            <div className='auto-container'>
              <div className='banner-content-title'>{title}</div>
              <div className='avatar-text-group'>
                <div className='user-profile-picture'>
                  <RoundedImage imageSource={userAvatarImage} alt='profile picture'/>
                  <VerticalMargin className='hide-for-medium' size={50} />
                </div>
                <HorizontalMargin className='show-for-medium' size={10} />
                <div className='article-username-rate-group'>
                  <div className='article-header-username'>{username}</div>
                  <div className ='article-rating-container'>
                    <div className='article-header-rating'>
                      <RatingStars rateNumber={(averageRating)} />
                    </div>
                    <div className='article-header-rated-times'>
                      {!ratingsCount ? 'No ratings yet' : `rated by ${ratingsCount} user${ratingsCount > 1 ? 's' : ''}`}
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div className='body-section'>
          <div className='auto-container wide-padding'>
            <div className='article-body'>
              <div dangerouslySetInnerHTML={{ __html: body }} />
            </div>
            <HorizontalLine />
            <div className='section-article-info'>  
              <div className='rate-impression'>
                <p>Rate this Article</p> <RatingStarsBox starClickHandle= {this.starClickHandle.bind(this)} /> 
              </div>  
              <div className='impressions-container'>
                <div className='impression'>
                 <LikeNumberGroup likeCount = {0} onClick={this.likeHandle.bind(this)}/>
                </div>
                <div className='impression'>
                <BookMarkNumberGroup bookMarkCount = {16} onClick ={this.bookmarkHandle.bind(this)}/>
                </div>
                <div className='impression'>
                <DisLikeNumberGroup dislikeCount = {0} onClick ={this.dislikeHandle.bind(this)} />
                </div>
              </div>
              <VerticalMargin className='show-for-medium' size={10} />
              <div>
                <div className='section-comment-create'>
                  <div>Leave comments</div>
                  <TextAreaInput className='article-comment-field' text='Your message' />
                  <button></button>
                </div>
                <div className='section-user-comments'>
                  <CommentEntries comments={comments} />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

SingleArticleView.propTypes = {
  articles: PropTypes.array.isRequired,
  match: PropTypes.object.isRequired,
  fetchArticle: PropTypes.func.isRequired,
  rateArticle: PropTypes.func.isRequired,
  getArticleAvgRating: PropTypes.func.isRequired,
  history: PropTypes.object,
  likeArticle: PropTypes.func,
  dislikeArticle: PropTypes.func,
  bookmarkArticle: PropTypes.func,
}

const mapStateToProps = (state) => ({
  articles: state.articles
});

const mapDispatchToProps = (dispatch) => ({
  fetchArticle: (id) => dispatch(getOneArticle(id)),
  rateArticle: (rating, articleId) => dispatch(rateArticle(rating, articleId)),
  getArticleAvgRating: (articleId) => dispatch(getArticleAvgRating(articleId)),
  likeArticle: (articleId) => dispatch(likeArticle(articleId)),
  dislikeArticle: (articleId) => dispatch(dislikeArticle(articleId)),
  bookmarkArticle: (articleId) => dispatch(bookmarkArticle(articleId))
});

export default connect(mapStateToProps, mapDispatchToProps)(SingleArticleView);



