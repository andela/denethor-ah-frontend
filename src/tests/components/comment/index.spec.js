import React from 'react';
import { mount } from 'enzyme';
import { CommentEntries, CreateComment}  from '../../../components/comments';
import { comments, commentEntry, count }  from '../../mock-data/comments';

const allCommentsImpressions = [];

describe('Test for the single comment component', () => {
	const props = {
		articleId: comments[0].articleId,
		addComment: jest.fn(),
		allCommentsImpressions
	};

	it('should render the comment section', () => {
		const wrapper = mount( <CreateComment {...props} /> );
		expect(wrapper.length).toBe(1);
		expect(wrapper.find('.comment-create-field').length).toBe(1);
		expect(wrapper.find('div').length).toBe(1);
		expect(wrapper.find('textarea').length).toBe(1);
		expect(wrapper.find('button').length).toBe(1);
		expect(wrapper).toMatchSnapshot();
	});

	it('should render the comment entry section', () => {
		const wrapper = mount(<CommentEntries allCommentsImpressions={count} comments={commentEntry}/>);
		expect(wrapper.length).toBe(1);
		expect(wrapper.find('.comment-entry').length).toBe(1);
	});

	it('should call toast error when server error occurs', () => {
		const addComment = jest.fn(() => Promise.reject({}));
		const props = {
			articleId: comments[0].articleId,
			addComment
		};
		const wrapper = mount( <CreateComment {...props} /> );
		wrapper.find('Button').at(0).simulate('click');
		expect(wrapper).toMatchSnapshot();
		expect(addComment).toHaveBeenCalled();
	});

	it('should show a toast when a validation error occurs', () => {
		const addComment = jest.fn(() => Promise.reject({ response: {
			status: 422
		}}));
		const props = {
			articleId: comments[0].articleId,
			addComment
		};
		const wrapper = mount( <CreateComment {...props} /> );
		wrapper.find('Button').at(0).simulate('click');
		expect(wrapper).toMatchSnapshot();
		expect(addComment).toHaveBeenCalled();
	});

	it('should show a toast when user is not logged in', () => {
		const addComment = jest.fn(() => Promise.reject({ response: {
			status: 401
		}}));
		const props = {
			articleId: comments[0].articleId,
			addComment
		};
		const wrapper = mount( <CreateComment {...props} /> );
		wrapper.find('Button').at(0).simulate('click');
		expect(wrapper).toMatchSnapshot();
		expect(addComment).toHaveBeenCalled();
	});
	it('should update state as user is typing', () => {
		const wrapper = mount( <CreateComment {...props} /> );
		wrapper.find('TextAreaInput').at(0).simulate('change', { target: { value: 'hi' } });
		expect(wrapper.state('commentBody')).toEqual('hi');
	});
});