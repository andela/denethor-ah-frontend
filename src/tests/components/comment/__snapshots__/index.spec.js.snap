// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for the single comment component should call toast error when server error occurs 1`] = `
<Comment
  addComment={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "articleId": "8ee5e8ed-ecdf-41c1-9b94-6c6bb712a77a",
            "commentBody": "",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": Promise {},
        },
      ],
    }
  }
  articleId="8ee5e8ed-ecdf-41c1-9b94-6c6bb712a77a"
>
  <div
    className="comment-create-field"
  >
    <TextAreaInput
      className="article-comment-field"
      onChange={[Function]}
      placeHolder="Your message"
      placeholder="Your message"
      value=""
    >
      <textarea
        className="textarea-input article-comment-field"
        onChange={[Function]}
        placeholder="Your message"
        value=""
      />
    </TextAreaInput>
    <Button
      className="article-comment-field__button"
      isRequestSent={false}
      onClick={[Function]}
      type="submit"
      value="submit"
    >
      <button
        className="article-comment-field__button"
        disabled={false}
        onClick={[Function]}
        type="submit"
      >
        submit
      </button>
    </Button>
  </div>
</Comment>
`;

exports[`Test for the single comment component should render the comment section 1`] = `
<Comment
  addComment={[MockFunction]}
  articleId="8ee5e8ed-ecdf-41c1-9b94-6c6bb712a77a"
>
  <div
    className="comment-create-field"
  >
    <TextAreaInput
      className="article-comment-field"
      onChange={[Function]}
      placeHolder="Your message"
      placeholder="Your message"
      value=""
    >
      <textarea
        className="textarea-input article-comment-field"
        onChange={[Function]}
        placeholder="Your message"
        value=""
      />
    </TextAreaInput>
    <Button
      className="article-comment-field__button"
      isRequestSent={false}
      onClick={[Function]}
      type="submit"
      value="submit"
    >
      <button
        className="article-comment-field__button"
        disabled={false}
        onClick={[Function]}
        type="submit"
      >
        submit
      </button>
    </Button>
  </div>
</Comment>
`;

exports[`Test for the single comment component should show a toast when a validation error occurs 1`] = `
<Comment
  addComment={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "articleId": "8ee5e8ed-ecdf-41c1-9b94-6c6bb712a77a",
            "commentBody": "",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": Promise {},
        },
      ],
    }
  }
  articleId="8ee5e8ed-ecdf-41c1-9b94-6c6bb712a77a"
>
  <div
    className="comment-create-field"
  >
    <TextAreaInput
      className="article-comment-field"
      onChange={[Function]}
      placeHolder="Your message"
      placeholder="Your message"
      value=""
    >
      <textarea
        className="textarea-input article-comment-field"
        onChange={[Function]}
        placeholder="Your message"
        value=""
      />
    </TextAreaInput>
    <Button
      className="article-comment-field__button"
      isRequestSent={false}
      onClick={[Function]}
      type="submit"
      value="submit"
    >
      <button
        className="article-comment-field__button"
        disabled={false}
        onClick={[Function]}
        type="submit"
      >
        submit
      </button>
    </Button>
  </div>
</Comment>
`;

exports[`Test for the single comment component should show a toast when user is not logged in 1`] = `
<Comment
  addComment={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "articleId": "8ee5e8ed-ecdf-41c1-9b94-6c6bb712a77a",
            "commentBody": "",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": Promise {},
        },
      ],
    }
  }
  articleId="8ee5e8ed-ecdf-41c1-9b94-6c6bb712a77a"
>
  <div
    className="comment-create-field"
  >
    <TextAreaInput
      className="article-comment-field"
      onChange={[Function]}
      placeHolder="Your message"
      placeholder="Your message"
      value=""
    >
      <textarea
        className="textarea-input article-comment-field"
        onChange={[Function]}
        placeholder="Your message"
        value=""
      />
    </TextAreaInput>
    <Button
      className="article-comment-field__button"
      isRequestSent={false}
      onClick={[Function]}
      type="submit"
      value="submit"
    >
      <button
        className="article-comment-field__button"
        disabled={false}
        onClick={[Function]}
        type="submit"
      >
        submit
      </button>
    </Button>
  </div>
</Comment>
`;
